### PR TITLE
things for number animation (so that vignesh verify this)

### PR DESCRIPTION
**Pull request by dcastroeyss**

### CODE REVIEW

File: mod/homepage/set/self/homepage_countup.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+include_set Abstract::CodeFile
```
Comment (srivig) [inline mod/homepage/set/self/homepage_countup.rb]: this should be script_countup.rb , I am fixing this

File: mod/homepage/template/self/homepage_numbers.haml [hunk 2]
```diff
@@ -4,10 +4,10 @@
   .mx-auto.mt-4
     .row
       .col-12.col-md-6
-        - categories.each do |cat|
+        - categories.each_with_index do |cat,i|
           .d-flex.mb-5.mr-4.justify-content-end
             .text-right.mx-3
-              %h1.font-weight-normal
+              %h1.font-weight-normal_count-ele{id: "count-ele-#{i}"}
```
Comment (srivig) [inline mod/homepage/template/self/homepage_numbers.haml]: missing `.` near `_count-ele`
Should be `%h1.font-weight-normal._count-ele{id: "count-ele-#{i}"}`

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 3]
```diff
@@ -107,8 +118,32 @@ isScrolledIntoView = (elem) ->
   elemBottom = elemTop + $(elem).height();
   ((elemBottom <= docViewBottom) && (elemTop >= docViewTop));
 
-animation = (elem) ->
-  # animation CountUp.js
+runAnimation = (elem) ->
+  options = { useEasing: true, useGrouping: true, separator: ',', decimal: '.', };
+  demo = new CountUp($(elem).attr('id'),  0, parseInt($(elem).text()), 0, 3.5, options);
```
Comment (srivig) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: why `demo` ?

Comment (srivig) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: we need to remove the delimiter `number_with_delimiter(number, delimiter: ",")` from `hompeage_numbers.rb' ` in order make this countUp js work properly.

File: mod/homepage/template/self/homepage_numbers.haml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+          .d-flex.mb-5.mr-4.justify-content-end
+            .text-right.mx-3
+              %h1.font-weight-normal_count-ele{id: "count-ele-#{i}"}
+                = cat.count
+              %h5.font-weight-normal.text-muted
+                = cat.title
+            .drop-shadow.rounded-circle.bg-white.image-box.small.m-2{class: "#{cat.color}-color"}
+              = icon_tag(cat.icon, class: "md-24 p-3")
+      .d-none.d-md-inline.col-5.mt-4
+        .carousel.carousel-fade.slide.drop-shadow{"data-ride" => "carousel"}
```
Comment (srivig) [inline mod/homepage/template/self/homepage_numbers.haml]: missing `.` near `_count-ele`
Should be `%h1.font-weight-normal._count-ele{id: "count-ele-#{i}"}`

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 5]
```diff
@@ -64,10 +107,48 @@ activateIntroTab = (tab)->
...
+isScrolledIntoView = (elem) ->
+  docViewTop = $(window).scrollTop();
+  docViewBottom = docViewTop + $(window).height();
+  elemTop = $(elem).offset().top;
+  elemBottom = elemTop + $(elem).height();
+  ((elemBottom <= docViewBottom) && (elemTop >= docViewTop));
+
+runAnimation = (elem) ->
+  options = { useEasing: true, useGrouping: true, separator: ',', decimal: '.', };
+  demo = new CountUp($(elem).attr('id'),  0, parseInt($(elem).text()), 0, 3.5, options);
```
Comment (srivig) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: we need to remove the delimiter `number_with_delimiter(number, delimiter: ",")` from `hompeage_numbers.rb' ` in order make this `parseInt($(elem).text())` work properly. number is already rendered with comma, so this parseInt won't work well. 
eg. 
output of parseInt("3434,5454")  will be 3434

Comment (srivig) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: also why `demo` ?

**Review by srivig — CHANGES_REQUESTED** [review]

File: db/migrate_cards/20180815220000_new_homepage_codenames.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# -*- encoding : utf-8 -*-
```
Comment (ethn) [inline db/migrate_cards/20180815220000_new_homepage_codenames.rb]: is this a new migration (and not a rename of the old?)

File: mod/homepage/set/self/script_countup.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+include_set Abstract::CodeFile
+Self::ScriptLibraries.add_item :script_countup
+
+def source_files
+  "vendor/CountUp.js/dist/countUp.js"
```
Comment (ethn) [inline mod/homepage/set/self/script_countup.rb]: `CountUp.js` is not an acceptable name for a directory.  our directories are lower case and do not have extensions.

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 8]
```diff
@@ -135,8 +149,32 @@ isScrolledIntoView = (elem) ->
...
+
+controlAnimate = (elem) ->
+  numberElementsControls.push( {id: $(elem).attr('id'), animated: false} )
+
+# has this element been animated?
+isAnimate = (id) ->
+  aux = false
+  numberElementsControls.forEach (elem) ->
+    if elem.id == id && elem.animated 
+      aux = true
```
Comment (ethn) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: does this actually return true?  It looks like it sets aux to true and then returns nothing.

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 9]
```diff
@@ -135,8 +149,32 @@ isScrolledIntoView = (elem) ->
...
+runAnimation = (elem) ->
+  options = { useEasing: true, useGrouping: true, separator: ',', decimal: '.', };
+  animationNumber = new CountUp($(elem).attr('id'),  0, parseInt($(elem).text()), 0, 3.5, options);
+  animationNumber.start()
+
+controlAnimate = (elem) ->
+  numberElementsControls.push( {id: $(elem).attr('id'), animated: false} )
+
+# has this element been animated?
+isAnimate = (id) ->
```
Comment (ethn) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: isAnimate**d**

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 10]
```diff
@@ -122,10 +135,11 @@ activateIntroTab = (tab)->
   active_panel.find('.carousel').carousel()
   active_panel.find('.carousel-item').first().addClass 'active'
 
-getNumbers = () -> 
+getNumberElements = () -> 
   values = []
-  $('.text-right.mx-3').each -> 
-    values.push( $(this).find('h1.font-weight-normal') )
+  $('h1.font-weight-normal._count-ele').each ->
```
Comment (ethn) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: `h1.font-weight-normal` is not a good selector.  It means the js will break if the style changes.

For javascript elements we typically prefer to use selectors that start with underscore. (and those selectors should NOT be used for styling)

File: mod/homepage/set/self/homepage_numbers.rb [hunk 11]
```diff
@@ -22,6 +22,5 @@ def featured_answers
 
   def count card_type
     number = nest card_type, view: :count
```
Comment (srivig) [inline mod/homepage/set/self/homepage_numbers.rb]: remove `number` as it is not used anymore

**Review by srivig — CHANGES_REQUESTED** [review]

File: mod/homepage/set/self/homepage_numbers.rb (hunk 11 above)
Comment (dcastroeyss) [reply mod/homepage/set/self/homepage_numbers.rb]: `def count card_type
    nest card_type, view: :count
 end`

Yes?

**Review by ethn — CHANGES_REQUESTED** [review]

File: db/migrate_cards/20180815220000_new_homepage_codenames.rb (hunk 6 above)
Comment (dcastroeyss) [reply db/migrate_cards/20180815220000_new_homepage_codenames.rb]: It's a name change

File: mod/homepage/set/self/script_countup.rb (hunk 7 above)
Comment (dcastroeyss) [reply mod/homepage/set/self/script_countup.rb]: when added the git submodule was added with that name (CountUp.js)

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 9 above)
Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: ok

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 8 above)
Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: when "aux" is true then "return" stops the "forEach" and then the value of "aux" is returned

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 10 above)
Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: `$('h1._count-ele')`
yes?

File: db/migrate_cards/20180815220000_new_homepage_codenames.rb (hunk 6 above)
Comment (ethn) [reply db/migrate_cards/20180815220000_new_homepage_codenames.rb]: Git doesn't think so.   Did you not delete the old file?

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 10 above)
Comment (ethn) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: Why not just `_count-ele`? Is `_count-ele` used anywhere that isn't supposed to be counted?

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 8 above)
Comment (ethn) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: ah, ok.  
This is so much cleaner in ruby ;)

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 12]
```diff
@@ -135,8 +149,32 @@ isScrolledIntoView = (elem) ->
...
+isAnimate = (id) ->
+  aux = false
+  numberElementsControls.forEach (elem) ->
+    if elem.id == id && elem.animated 
+      aux = true 
+      return
+  aux  
+
+# this element is already animated now
+nowIsAnimate = (id) ->
```
Comment (ethn) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: wait.  what's the difference between this and isAnimated?

maybe the other should be `isAnimatable` and this is `isAnimated`?

sorry, I was confused before.

File: mod/homepage/set/self/script_countup.rb (hunk 7 above)
Comment (ethn) [reply mod/homepage/set/self/script_countup.rb]: I know.  please fix it.

File: db/migrate_cards/20180815220000_new_homepage_codenames.rb (hunk 6 above)
Comment (dcastroeyss) [reply db/migrate_cards/20180815220000_new_homepage_codenames.rb]: No, I change the name

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 12 above)
Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: this function only sets the element as animated

Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: isAnimated is to know if an element has already been animated

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 10 above)
Comment (dcastroeyss) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: ok

File: mod/homepage/set/self/script_countup.rb (hunk 7 above)
Comment (dcastroeyss) [reply mod/homepage/set/self/script_countup.rb]: ok

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee (hunk 12 above)
Comment (ethn) [reply mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: oh, now that I'm looking at the functionality, I see that it's not a boolean call.  (And I should note that even if it returned true/false, usually a boolean call should not have side effects)

this should be something like `animateElem`, no

File: mod/homepage/lib/javascript/script_homepage_carousel.js.coffee [hunk 13]
```diff
@@ -135,8 +149,32 @@ isScrolledIntoView = (elem) ->
...
+# has this element been animated?
+isAnimated = (id) ->
+  aux = false
+  numberElementsControls.forEach (elem) ->
+    if elem.id == id && elem.animated 
+      aux = true 
+      return
+  aux  
+
+# this element is already animated now
```
Comment (ethn) [inline mod/homepage/lib/javascript/script_homepage_carousel.js.coffee]: # animate this element

File: mod/homepage/vendor/CountUp/.editorconfig [hunk 14]
```diff
@@ -0,0 +1,12 @@
+# This file is for unifying the coding style for different editors and IDEs
```
Comment (ethn) [inline mod/homepage/vendor/CountUp/.editorconfig]: now this is not a submodule.

You need to make it a submodule with the correct directory name.  you can control the directory name. read about the `git submodule` command.

File: .gitmodules [hunk 15]
```diff
@@ -17,6 +17,6 @@
 [submodule "mod/homepage/vendor/wodry"]
 	path = mod/homepage/vendor/wodry
 	url = https://github.com/srivig/wodry.git
-[submodule "mod/homepage/vendor/CountUp.js"]
-	path = mod/homepage/vendor/CountUp.js
-	url = https://github.com/inorganik/CountUp.js
+[submodule "mod/homepage/vendor/CountUp"]
+	path = mod/homepage/vendor/CountUp
+	url = https://github.com/inorganik/countUp.js.git
```
Comment (ethn) [inline .gitmodules]: we control the name of the submodule and the path (and we want them to be lowercase).  we _don't_ control their url.

**Pull request merged**